edx_rbac/mixins.py

```diff
@@ -61,3 +64,128 @@ def check_permissions(self, request):
             self.permission_denied(
                 request,
                 message=('MISSING: {}'.format(', '.join(missing_permissions))))
+
+
+class PermissionRequiredForListingMixin(PermissionRequiredMixin):
+    """
+    Mixin for checking role-based access against multiple objects, specifically
+    using the "list" action of a DRF ViewSet.  The brunt of the work is done in
+    the `accessible_contexts` cached property - this property determines the set
+    of "contexts" (probably identifiers of some model) to which the requesting
+    user has access under any of the `allowed_roles`.  It's pretty opinionated and
+    you _must_ define the following fields:
+
+    `list_lookup_field` - The field of the model you want to list instances of
+    that holds values defined by `accessible_contexts`.
+
+    `allowed_roles` - A list of one or more roles against which access is checked.  It's disjunctive,
+    which means that if "role_1" grants access to resources "A" and "B", and
+    "role_2" grants access to resource "C", defining `allowed_roles = ['role_1', 'role_2']`
+    will cause resources "A", "B", and "C" to be returned by a listing operation
+    for a user who is granted both "role_1" and "role_2".
+
+    `role_assignment_class` - The RoleAssignmentClass against which DB-defined access is checked.
+
+    `base_queryset` (property) - A queryset which acts as the "base case".  It should generally return
+    all accessible instances for a user who has access to anything within this viewset (like a superuser
+    or admin).
+    """
+    # This flag indicates whether requesting users with `is_staff = True`
+    # should ever encounter a PermissionDenied exception.
+    staff_are_never_forbidden = True
+
+    # This flag indicates whether requesting users with `is_superuser = True`
+    # should always have the `ALL_ACCESS_CONTEXT` included in their set of accessible contexts.
+    superusers_can_access_anything = True
+
+    # The field of the model you want to list instances of
+    # that holds values defined by `accessible_contexts`.
+    list_lookup_field = None
+
+    # A list of one or more roles against which access is checked.  It's disjunctive,
+    # which means that if "role_1" grants access to resources "A" and "B", and
+    # "role_2" grants access to resource "C", defining `allowed_roles = ['role_1', 'role_2']`
+    # will cause resources "A", "B", and "C" to be returned by a listing operation
+    # for a user who is granted both "role_1" and "role_2".
+    allowed_roles = []
+
+    # The RoleAssignmentClass against which DB-defined access is checked
```

Comment:
Nit: For all of these it feels superfluous to have the same comment here and in the docstring, but I'm already over it as I finish typing this